internal/clients/aws.go

```diff
@@ -7,61 +7,160 @@ package clients
 import (
 	"context"
 
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"k8s.io/apimachinery/pkg/types"
+
 	"github.com/crossplane/crossplane-runtime/pkg/resource"
 	"github.com/pkg/errors"
+	"github.com/upbound/upjet/pkg/terraform"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
-	"github.com/upbound/upjet/pkg/terraform"
+	"github.com/upbound/provider-aws/apis/v1beta1"
 )
 
 const (
 	// Terraform provider configuration keys for AWS credentials.
-	keyRegion          = "region"
-	keyAccountId       = "account_id"
-	keySessionToken    = "token"
-	keyAccessKeyID     = "access_key"
-	keySecretAccessKey = "secret_key"
+	keyRegion                    = "region"
+	keyAccountId                 = "account_id"
+	keySessionToken              = "token"
+	keyAccessKeyID               = "access_key"
+	keySecretAccessKey           = "secret_key"
+	keyAssumeRoleWithWebIdentity = "assume_role_with_web_identity"
+	keyWebIdentityTokenFile      = "web_identity_token_file"
+	keyAssumeRole                = "assume_role"
+	keyTags                      = "tags"
+	keyTransitiveTagKeys         = "transitive_tag_keys"
 )
 
-// TerraformSetupBuilder returns Terraform setup with provider specific
-// configuration like provider credentials used to connect to cloud APIs in the
-// expected form of a Terraform provider.
-func TerraformSetupBuilder(version, providerSource, providerVersion string) terraform.SetupFn { //nolint:gocyclo
-	return func(ctx context.Context, client client.Client, mg resource.Managed) (terraform.Setup, error) {
-		cfg, err := GetAWSConfig(ctx, client, mg)
+func SelectTerraformSetup(version, providerSource, providerVersion string) terraform.SetupFn {
+	return func(ctx context.Context, c client.Client, mg resource.Managed) (terraform.Setup, error) {
+		pc := &v1beta1.ProviderConfig{}
+		if err := c.Get(ctx, types.NamespacedName{Name: mg.GetProviderConfigReference().Name}, pc); err != nil {
+			return terraform.Setup{}, errors.Wrap(err, "cannot get referenced Provider")
+		}
+		ps := terraform.Setup{
+			Version: version,
+			Requirement: terraform.ProviderRequirement{
+				Source:  providerSource,
+				Version: providerVersion,
+			},
+		}
+		account, err := getAccountId(ctx, c, mg)
 		if err != nil {
-			return terraform.Setup{}, errors.Wrap(err, "cannot get AWS config")
+			return terraform.Setup{}, errors.Wrap(err, "cannot get account id")
+		}
+		ps.ClientMetadata = map[string]string{
+			keyAccountId: account,
 		}
-		if cfg.Region == "" && mg.GetObjectKind().GroupVersionKind().Group == "iam.aws.upbound.io" {
-			cfg.Region = "us-east-1"
+
+		var err2 error
+		if (len(pc.Spec.AssumeRoleChain) > 1 || pc.Spec.Endpoint != nil) && (pc.Spec.Credentials.Source == authKeyIRSA || pc.Spec.Credentials.Source == authKeyWebIdentity || pc.Spec.Credentials.Source == authKeyUpbound) {
```

Comment:
We should be invoking the default path if we have an AssumeRole chain of length greater than 1 or an endpoint configuration independent of the configured authentication scheme.